tests/_async/test_connection.py

```diff
@@ -141,6 +141,56 @@ async def connect_tcp(
         assert response.content == b"Request body exceeded 1,000,000 bytes"
 
 
+@pytest.mark.filterwarnings("ignore::pytest.PytestUnraisableExceptionWarning")
+@pytest.mark.anyio
+async def test_write_error_with_response_sent_http2():
+    """
+    If a server half-closes the connection while the client is sending
+    the request, it may still send a response. In this case the client
+    should successfully read and return the response.
+
+    See also the `test_write_error_without_response_sent_http2` test above.
+    """
+
+    origin = Origin(b"https", b"example.com", 443)
+    network_backend = AsyncMockBackend(
+        [
+            hyperframe.frame.SettingsFrame().serialize(),
+            hyperframe.frame.WindowUpdateFrame(
+                stream_id=0, window_increment=2147418112
+            ).serialize()
+            + hyperframe.frame.WindowUpdateFrame(
+                stream_id=1, window_increment=2147418112
+            ).serialize()
+            + hyperframe.frame.HeadersFrame(
+                stream_id=1,
+                data=hpack.Encoder().encode(
+                    [
+                        (b":status", b"413"),
+                        (b"content-type", b"plain/text"),
+                    ]
+                ),
+                flags=["END_HEADERS"],
+            ).serialize()
+            + hyperframe.frame.DataFrame(
+                stream_id=1,
+                data=b"Request body exceeded 1,000,000 bytes",
+                flags=["END_STREAM"],
+            ).serialize()
+            + hyperframe.frame.RstStreamFrame(stream_id=1, error_code=0).serialize(),
```

Comment:
Interesting. Is it intentional/neccessary-for-the-test that the data for these frames is concatenated?

Reply:
I ran into some issues without frame concatenation.